### ISSUE TITLE
Enhancement: truncate every class implementing Versioned 

### ISSUE THREAD

**Issue by DrMartinGonzo** [axllent/silverstripe-version-truncator#11]
Hi, and thanks for the module, much cleaner db before deploying website !
Would it be possible to extend this module to safely truncate versions of not only SiteTree but also every class implementing Versioned ?

**Comment by axllent** [comment]
Hi there. Yes theoretically that should be possible (and I have thought about it before). The problem is I would need to rewrite the whole class to do that. I will leave this issue open because it is on my list of things to do... one day... when I have time... :)

Ideally speaking it would:
- Attach to every class with Versioned
- Be configurable (enough) to specify defaults per class (eg: you may want 10 SiteTree versions, but only 2 Files)
- Allowed to be turned off for a certain class that implemented versioned

Thanks for your feedback!

**Comment by robingram** [comment]
I'd be interested to see how this might apply to files/images if it is extended to cover all Versioned objects.

**Comment by axllent** [comment]
So would I to be honest! It's all theoretical at the moment, and the issue is I don't have time to rewrite and test this (which would need extensive testing). Since +- SS 4.2 there is a lot of underlying linkage between all these objects as well.

**Comment by tractorcow** [comment]
I did this for my own module.

https://github.com/isobar-nz/silverstripe-versionprune/blob/master/src/ClearArchivedRecords.php#L101-L108

**Comment by axllent** [comment]
Thanks @tractorcow - that's a neat way to do it (and I've never used `yield` before) :+1:

**Comment by axllent** [comment]
OK guys, it's been a busy weekend. I have completely rewritten the module and released a 3.0.0-beta1 pre-release (`composer require axllent/silverstripe-version-truncator:^3.0`) which I would love some feedback & testing on.

It now supports ALL versioned DataObjects, with different (and editable) configurations for per class. This module is not configurable with previous configurations due to the changes, so please read the [README](https://github.com/axllent/silverstripe-version-truncator/blob/3.0.0-beta1/README.md) to see more of how it works now, and the generic defaults for SiteTree & other DataObjects.

The main differences between the 3.* and the 2.* versions are:
- Automatic pruning is now only done on `onAfterPublish()`, and not every draft write. This means that a stack of drafts for any DataObject can build up prior to a publish. Once that DataObject is published however, then it is pruned accordingly.
- SiteTree now has a default of 10 versions and 0 drafts, meaning that all drafts are expunged once a page is published. SS4 seems to have some major changes that store an additional draft for every publish (rather than transforming that draft into a published version), so keeping say 5 drafts isn't a reliable way of necessarily keeping 5 actual drafts, but often ends up being identical copies of what got published immediately afterwards.
- All other DataObjects only keep (by default) the latest published version and no drafts. This includes Files etc. I figured that anyone using this module doesn't want that additional historical crud in their database anyway, however does want the flexibility of versioning for Draft / Modified abilities.
- There is one new thing I noticed (not sure f this is new or not in later SS4 versions), however SS does not delete any records for deleted files. This appears to be related to the fact that there are some post-deletion database actions that are run regarding the file's internal linking. I have added a separate task to deal with deleted files, which deleted all versioned records for those.

Feedback very welcome!

**Comment by tractorcow** [comment]
ss4's major change is:

 - a save in draft saves only to draft
 - a save in live saves to both stages (whereas in ss3, it would ONLY save to live).

@axllent make sure that you keep the latest stage version, otherwise versioning in the CMS will get confused and think the record is archived (and not on draft).

I.e. don't delete the version matching the BaseTable.Version number. :)

When I worked on a version truncator fix for another site, I had to explicitly exclude deleting any version that had as specific version in any table.

In other words, if you have NO live versions, you'd need to keep at least 1 draft version. If you have at least one live version, then that is also the draft version too. :)

**Comment by axllent** [comment]
@tractorcow Thanks for the info! That explains it (the SS4 Versiond publish change) - I obviously hadn't noticed this was the case until recently. It really doesn't make any sense to me why the system would keep a 1:1 draft duplicate of every published version too, but that's a different topic entirely, and one this version of the module now caters for by default ;-)

As far as the other points go (about keeling the latest version for non-published DOs), this module shouldn't be affected by that as it only runs on `onAfterPublish()`. I completely understand the issue you are pointing out though, v2 of my module had to cater for that scenario too (after I learned the hard way). I couldn't see the point in maintaining this functionality on every draft save too, just extra DB queries that gave very little benefit and lots of headaches (hence why now `onAfterPublish()` only).

**Comment by axllent** [comment]
Version 3.0.0 has been released which implements truncating on all versioned objects. If you experience any issues then please open a new ticket, thanks.
